vendor_skins/Evercast/libobs-d3d11/d3d11-texture2d.cpp

```diff
@@ -0,0 +1,313 @@
+					     (void **)&dxgi_res);
+		if (FAILED(hr)) {
+			blog(LOG_WARNING,
+			     "InitTexture: Failed to query "
+			     "interface: %08lX",
+			     hr);
+		} else {
+			GetSharedHandle(dxgi_res);
+
+			if (flags & GS_SHARED_KM_TEX) {
+				ComPtr<IDXGIKeyedMutex> km;
+				hr = texture->QueryInterface(
+					__uuidof(IDXGIKeyedMutex),
+					(void **)&km);
+				if (FAILED(hr)) {
+					throw HRError("Failed to query "
+						      "IDXGIKeyedMutex",
+						      hr);
+				}
+
+				km->AcquireSync(0, INFINITE);
+				acquired = true;
+			}
+		}
+	}
+}
+
+void gs_texture_2d::InitResourceView()
+{
+	HRESULT hr;
+
+	memset(&resourceDesc, 0, sizeof(resourceDesc));
+	resourceDesc.Format = dxgiFormat;
+
+	if (type == GS_TEXTURE_CUBE) {
+		resourceDesc.ViewDimension = D3D11_SRV_DIMENSION_TEXTURECUBE;
+		resourceDesc.TextureCube.MipLevels =
+			genMipmaps || !levels ? -1 : levels;
+	} else {
+		resourceDesc.ViewDimension = D3D11_SRV_DIMENSION_TEXTURE2D;
+		resourceDesc.Texture2D.MipLevels =
+			genMipmaps || !levels ? -1 : levels;
+	}
+
+	hr = device->device->CreateShaderResourceView(texture, &resourceDesc,
+						      shaderRes.Assign());
+	if (FAILED(hr))
+		throw HRError("Failed to create resource view", hr);
+}
+
+void gs_texture_2d::InitRenderTargets()
+{
+	HRESULT hr;
+	if (type == GS_TEXTURE_2D) {
+		D3D11_RENDER_TARGET_VIEW_DESC rtv;
+		rtv.Format = dxgiFormat;
+		rtv.ViewDimension = D3D11_RTV_DIMENSION_TEXTURE2D;
+		rtv.Texture2D.MipSlice = 0;
+
+		hr = device->device->CreateRenderTargetView(
+			texture, &rtv, renderTarget[0].Assign());
+		if (FAILED(hr))
+			throw HRError("Failed to create render target view",
+				      hr);
+	} else {
+		D3D11_RENDER_TARGET_VIEW_DESC rtv;
+		rtv.Format = dxgiFormat;
+		rtv.ViewDimension = D3D11_RTV_DIMENSION_TEXTURE2DARRAY;
+		rtv.Texture2DArray.MipSlice = 0;
+		rtv.Texture2DArray.ArraySize = 1;
+
+		for (UINT i = 0; i < 6; i++) {
+			rtv.Texture2DArray.FirstArraySlice = i;
+			hr = device->device->CreateRenderTargetView(
+				texture, &rtv, renderTarget[i].Assign());
+			if (FAILED(hr))
+				throw HRError("Failed to create cube render "
+					      "target view",
+					      hr);
+		}
+	}
+}
+
+#define SHARED_FLAGS (GS_SHARED_TEX | GS_SHARED_KM_TEX)
+
+gs_texture_2d::gs_texture_2d(gs_device_t *device, uint32_t width,
+			     uint32_t height, gs_color_format colorFormat,
+			     uint32_t levels, const uint8_t *const *data,
+			     uint32_t flags_, gs_texture_type type,
+			     bool gdiCompatible, bool nv12_)
+	: gs_texture(device, gs_type::gs_texture_2d, type, levels, colorFormat),
+	  width(width),
+	  height(height),
+	  flags(flags_),
+	  dxgiFormat(ConvertGSTextureFormat(format)),
+	  isRenderTarget((flags_ & GS_RENDER_TARGET) != 0),
+	  isGDICompatible(gdiCompatible),
+	  isDynamic((flags_ & GS_DYNAMIC) != 0),
+	  isShared((flags_ & SHARED_FLAGS) != 0),
+	  genMipmaps((flags_ & GS_BUILD_MIPMAPS) != 0),
+	  sharedHandle(GS_INVALID_HANDLE),
+	  nv12(nv12_)
```

Comment:
I wonder if nv12 is at play during capture or only during encoding.  I think it's tied directly ino the rtmp / h264 layer and not into the webrtc encoders.